                                         Case Study: Problem-1 [ 3-4 Hours]
A shop-keeper would like to decide weekly pricing of their products. He/She generally plans pricing for each quarter and would like to achieve following:
1. Clearout at least 60% of total inventory.
2. achieve total margin across all products (i.e. selling price - cost price).
Following are a few things to consider while deciding pricing of products.
3. There shouldn’t be much difference in the pricing from one week to another.
4. A particular price selected applies for all the products in the group.
                                                    Expectations
1. Develop a mathematical model that can help shop-keeper to identify their pricing for a quarter (13 weeks in the provided example)
2. Model has to suggest a pricing (price-point) that will be applied across all the products to achieve highest possible margin considering inventory clearance of 60% (total inventory in the provided example is 264443.0 units, 60% of this has to be sold)
3. Max Price change from one week to next is 20 units
4. Shop-keeper has to cater posed weekly demand to customers if inventory is available
5. A single price-point for a week applies to all the products as they belong to same group
6. Implementation of model in python and obtain some initial results
7. Algorithmic details to solve the model and provide insights on complexity of model in scaling to larger number of products/weeks.

# Data Cleaning and Summary

In [1]:
import pandas as pd
import numpy as np
from collections import defaultdict

In [2]:
data=pd.read_csv('/Users/Z003CNF/Desktop/AssignmentData.csv')

In [27]:
## check size of data and nulls if any 
data.shape

(1265745, 8)

In [10]:
data['product_id'].nunique()

6491

In [28]:
1265745/6491 # one product has 195 rows 

195.0

In [23]:
d1['price'].nunique()

15

In [29]:
13*15    # for each prod in each week 15 differnet price points
          #are there of which we need to select one price point to maximize total margin 
          #and demand is atleast 60% of inv

195

In [34]:
a=data['product_id'].value_counts() # to confirm if total 195 rows for each prod.
a.nunique()

1

In [36]:
# Handling missing values if any

# Exploratory Data Analysis

In [6]:
# Total Inventory for all prods at the start of quater

prod_inv=data.groupby(['product_id']).agg(inventory=pd.NamedAgg(column="total_inventory", aggfunc="mean")).reset_index()
prod_inv['inventory'].sum()


264443.0

In [44]:
0.6*264443 # This much inventory needs to be cleared

158665.8

In [65]:
# check if sp is always > cp for all prod's and possible price points
data['is_sp_gt_cp']=np.where(data['selling_price'] >= data['cost_price'],1,0)
data['is_sp_gt_cp'].sum()

1265745

In [ ]:
## Selling price was always > cost price hence we will always have positive margin.

In [56]:
# check if total quater demand is less than inv. available at start of quater
# taken max of a demand over all options in a week
prod_week_demand=data.groupby(['product_id','week']).agg(Demand=pd.NamedAgg(column="demand", aggfunc="max"))
prod_demand=prod_week_demand.groupby(['product_id']).agg(Tot_Demand=pd.NamedAgg(column="Demand", aggfunc="sum"))
prod_demand.sum()

Tot_Demand    620285.0
dtype: float64

In [61]:
# Compare inventory and demand
# maximum demand scenerio is 0.6 M but overall inventory is 0.26 M , so if all the products have high demand scenerio
# we may not be able to fulfill all of the demand.

prods=pd.merge(prod_inv,prod_demand, on=["product_id"])
prods['if']=np.where(prods['inventory'] >= prods['Tot_Demand'],1,0)
prods['if'].sum()

1697

In [62]:
#of 6k products, we have enough invnetory for 1.6 k products, if we consider maximum demand scenerios

### Objective Function

#####  $ Maximize: demand*(sp*x[i,j,k]-cp*x[i,j,k]) $

### Requirement 1: Inventory Clearence
##### Demand will vary as per the price we set and it will be met/un-met based on available inventory
##### $ total demand >= 158665.8 $
##### and total demand at particular price point if selected (for whole quater) <= inv. available start of quater

### Requirement 2: WOW price variation
##### $ For all i : abs(X[i,j,k]-X[i,j+1,k]) < 20    # it will be a non-linear constraint so will linearize it by added two constraints <20 & >20 $

# Optimization Model 

###### Find the optimum prices for products based on above two requirements

In [3]:
data['margin']=data['selling_price']-data['cost_price']

In [4]:
# test data for 2 products
products=data['product_id'].unique()
weeks=[202136, 202142, 202137, 202139, 202146, 202138, 202140, 202141,202147, 202143, 202144, 202145, 202148]
prices=[ 5.0, 10.0, 55.0, 60.0, 65.0, 70.0, 75.0, 15.0, 20.0, 25.0, 30.0, 35.0, 40.0, 45.0, 50.0]

combos = [(p, w, a) for p in products for w in weeks for a in prices]

In [7]:
#data1=data.loc[data['product_id'].isin(products)]
dict_demand=data.set_index(['product_id','week','price']).demand.to_dict()
dict_margin=data.set_index(['product_id','week','price']).margin.to_dict()
dict_inv_quater=prod_inv.set_index(['product_id']).inventory.to_dict()

In [8]:
from pulp import *
prob = LpProblem("Pricing_Problem",LpMaximize)
dvar = LpVariable.dicts("price_points",combos, cat="Binary")


In [9]:
## objective function:

prob += (
    lpSum([dvar[i] * dict_margin[i] * dict_demand[i] for i in combos])
)


In [10]:
## Constraint 1 --- Only. one price point is to be chosen for a product in a week

for p in products:
    for w in weeks:
        prob += (
            lpSum([dvar[(p,w,a)] for a in prices]) == 1
        )


In [333]:
## Constraint 2 ---- Demand to be met if enough inventory is available at start of quater.

for p in products:
    prob += (
        lpSum([dvar[(p,w,a)] for w in weeks for a in prices]) <= dict_inv_quater[p]
    )
    

In [316]:
## Constraint 3 ---- Demand should be as much as to clear 60% inventory

prob += (
    lpSum([dvar[i] * dict_demand[i] for i in combos]) >= 21
)


In [ ]:
## Constraint 4 ---- The price change from a week to next < 20 

prob += (
    lpSum(dvar[(p, w, j)] for w in weeks for j in prices)- lpSum(dvar[(p, w+1, j)] for w in weeks for j in prices) < 20
    
    
    

In [334]:
prob.writeLP("pricing.lp")

# The problem is solved using PuLP's choice of Solver
prob.solve()

# The status of the solution is printed to the screen
print("Status:", LpStatus[prob.status])

#for v in prob.variables():
    #print(v.name, "=", v.varValue)

Welcome to the CBC MILP Solver 
Version: 2.10.3 
Build Date: Dec 15 2019 

command line - /Users/Z003CNF/opt/anaconda3/lib/python3.9/site-packages/pulp/apis/../solverdir/cbc/osx/64/cbc /var/folders/my/1126m1255k36stt7f2r98h_w0000gq/T/4a6061c88a6348bf977666cda00ca51b-pulp.mps max timeMode elapsed branch printingOptions all solution /var/folders/my/1126m1255k36stt7f2r98h_w0000gq/T/4a6061c88a6348bf977666cda00ca51b-pulp.sol (default strategy 1)
At line 2 NAME          MODEL
At line 3 ROWS
At line 90879 COLUMNS
At line 5999853 RHS
At line 6090728 BOUNDS
At line 7356474 ENDATA
Problem MODEL has 90874 rows, 1265745 columns and 2531490 elements
Coin0008I MODEL read with 0 errors
Option for timeMode changed from cpu to elapsed
Problem is infeasible - 3.11 seconds
Option for printingOptions changed from normal to all
Total time (CPU seconds):       6.57   (Wallclock seconds):       7.55

Status: Infeasible
